detector/gpg.go

```diff
@@ -27,64 +28,52 @@ func WatchGPG(gpgPubringPath string, requestGPGCheck chan bool) {
 	initWatcher()
 	defer notify.Stop(events)
 
-	for {
-		select {
-		case event := <-events:
-			switch event.Event() {
-			case notify.InOpen:
-				select {
-				case requestGPGCheck <- true:
-				default:
-				}
+	for event := range events {
+		switch event.Event() {
+		case notify.InOpen:
+			select {
+			case requestGPGCheck <- true:
 			default:
-				log.Debugf("pubring.kbx received file event '%+v', recreating the watcher.", event.Event())
-				notify.Stop(events)
-				time.Sleep(5 * time.Second)
-				initWatcher()
 			}
+		default:
+			log.Debugf("pubring.kbx received file event '%+v', recreating the watcher.", event.Event())
+			notify.Stop(events)
+			time.Sleep(5 * time.Second)
+			initWatcher()
 		}
 	}
 }
 
 // CheckGPGOnRequest checks whether YubiKey is actually waiting for a touch on a GPG request
-func CheckGPGOnRequest(requestGPGCheck chan bool, notifiers *sync.Map) {
-	for {
-		<-requestGPGCheck
+func CheckGPGOnRequest(requestGPGCheck chan bool, notifiers *sync.Map, ctx *gpgme.Context) {
+	check := func(response chan error, ctx *gpgme.Context, t *time.Timer) {
+		err := ctx.AssuanSend("LEARN", nil, nil, func(status, args string) error {
+			log.Debugf("AssuanSend/status: %v, %v", status, args)
 
-		for i := 0; i < 20; i++ {
-			if isGPGCardBusy() {
-				notifiers.Range(func(k, v interface{}) bool {
-					v.(chan notifier.Message) <- notifier.GPG_ON
-					return true
-				})
-
-				for isGPGCardBusy() || isGPGCardBusy() {
-					// wait...
-				}
+			return nil
+		})
+		t.Stop()
+		response <- err
```

Comment:
```suggestion
		if !t.Stop() {
			response <- err
		}
```

If you stop the timer, there will be nobody to consume the message from `response`, so this send will block forever.